proxy/src/opentsdb/types.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::collections::{HashMap, HashSet};
+
+use bytes::Bytes;
+use ceresdbproto::storage::{
+    value, Field, FieldGroup, Tag, Value, WriteSeriesEntry, WriteTableRequest,
+};
+use common_util::error::BoxError;
+use itertools::Itertools;
+use serde::Deserialize;
+use serde_json::from_slice;
+use snafu::ResultExt;
+
+use crate::error::{Internal, InternalNoCause, Result};
+
+const OPENTSDB_DEFAULT_FIELD: &str = "value";
+
+#[derive(Debug)]
+pub struct PutRequest {
+    pub points: Bytes,
+
+    pub summary: bool,
+    pub details: bool,
+    pub sync: bool,
+    pub sync_timeout: i32,
+}
+
+impl PutRequest {
+    pub fn new(points: Bytes, params: PutParams) -> Self {
+        PutRequest {
+            points,
+            summary: params.summary,
+            details: params.details,
+            sync: params.sync,
+            sync_timeout: params.sync_timeout,
+        }
+    }
+}
+
+pub type PutResponse = ();
+
+/// Query string parameters for put api
+///
+/// It's derived from query string parameters of put described in
+/// doc of OpenTSDB 2.4:
+///     http://opentsdb.net/docs/build/html/api_http/put.html#requests
+///
+/// NOTE:
+///     - all the params is unimplemented.
+#[derive(Debug, Default, Deserialize)]
+#[serde(default)]
+pub struct PutParams {
+    // TODO: How to represent a `Present` param ?
+    // now use bool to represent it, so the URL must be "?summary=true/false"
+    // but the `Present` datatype means "?summary" is ok.
+    pub summary: bool,
+    pub details: bool,
+    pub sync: bool,
+    pub sync_timeout: i32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Point {
+    pub metric: String,
+    pub timestamp: u64,
+    // TODO: OpenTSDB's value support Integer, Float, String. How to represent it and parse from json ?
+    pub value: f64,
```

Comment:
You could try defining the `value` as an enum.

Reply:
This link may help: https://serde.rs/enum-representations.html#untagged

Reply:
fixed.